Core/Overlay/WebsocketBroadcastServer.cs

```diff
@@ -0,0 +1,210 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Core.Overlay
+{
+    public interface IBroadcastServer
+    {
+        public Task Send(string message, CancellationToken cancellationToken);
+    }
+
+    /// A websocket server capable of accepting new connections and broadcasting messages.
+    /// It cannot receive messages.
+    public class WebsocketBroadcastServer : IBroadcastServer, IAsyncDisposable
+    {
+        private readonly List<Connection> _connections = new List<Connection>();
+        private HttpListener? _httpListener;
+
+        private struct Connection : IAsyncDisposable
+        {
+            internal IPEndPoint RemoteEndPoint { get; init; }
+            internal WebSocket WebSocket { get; init; }
+            internal Task ReaderTask { get; init; }
+            internal CancellationTokenSource TokenSource { get; init; }
+
+            public async ValueTask DisposeAsync()
+            {
+                await ReaderTask;
+                WebSocket.Dispose();
+            }
+        }
+
+        private readonly string _host;
+        private readonly int _port;
+        private readonly ILogger<WebsocketBroadcastServer> _logger;
+        private readonly SemaphoreSlim _connectionsSemaphore = new SemaphoreSlim(initialCount: 100, maxCount: 100);
+
+        public WebsocketBroadcastServer(ILogger<WebsocketBroadcastServer> logger, string host, int port)
+        {
+            _logger = logger;
+            _host = host;
+            _port = port;
+        }
+
+        /// Send a message to all currently connected clients.
+        public async Task Send(string message, CancellationToken cancellationToken)
+        {
+            byte[] bytes = Encoding.UTF8.GetBytes(message);
+            await _connectionsSemaphore.WaitAsync(cancellationToken);
+            try
+            {
+                foreach (var connection in _connections)
+                {
+                    await connection.WebSocket.SendAsync(bytes, WebSocketMessageType.Text, endOfMessage: true,
```

Comment:
don't think that this should end if one client has an error, ideally it sends to all that it can, skipping ones with errors (try should be inside the foreach?)